src/Functions/FunctionsProgrammingClassification.cpp

```diff
@@ -0,0 +1,199 @@
+#include <Functions/FunctionsTextClassification.h>
+#include <Common/FrequencyHolder.h>
+#include <Functions/FunctionFactory.h>
+#include <IO/ReadHelpers.h>
+
+#include <unordered_map>
+
+namespace DB
+{
+/**
+  * Determine the programming language from the source code.
+  * We calculate all the unigrams and bigrams of commands in the source code.
+  * Then using a marked-up dictionary with weights of unigrams and bigrams of commands for various programming languages
+  * Find the biggest weight of the programming language and return it
+  */
+struct ProgrammingClassificationImpl
+{
+
+    using ResultType = String;
+    /// Calculate total weight
+    static ALWAYS_INLINE inline Float64 stateMachine(const FrequencyHolder::Map & standard, std::unordered_map<String, Float64> & model)
+    {
+        Float64 res = 0;
+        for (auto & el : model)
+        {
+            /// Try to find each n-gram in dictionary
+            auto it = standard.find(el.first);
+            if (it != standard.end())
+            {
+                res += el.second * it->getMapped();
+            }
+        }
+        return res;
+    }
+
+
+    static void constant(String data, String & res)
+    {
+        auto & programming_freq = FrequencyHolder::getInstance().getProgrammingFrequency();
+        std::unordered_map<String, Float64> data_freq;
+
+        String prev_command;
+        String command;
+        /// Select all commands from the string
+        for (size_t i = 0; i < data.size();)
+        {
+            /// Assume that all commands are split by spaces
+            if (!isspace(data[i]))
+            {
+                command.push_back(data[i]);
+                ++i;
+
+                while ((i < data.size()) && (!isspace(data[i])))
+                {
+                    command.push_back(data[i]);
+                    ++i;
+                }
+                if (prev_command == "")
+                {
+                    prev_command = command;
+                }
+                else
+                {
+                    data_freq[prev_command + command] += 1;
```

Comment:
Do we really need to concat strings?

Reply:
In the dictionary, we have unigrams and bigrams (words and concatenated word pairs) so we need to, yes